renter/hostdb/weight.go

```diff
@@ -0,0 +1,465 @@
+	}
+
+	// Compute the uptime ratio, but shift by 0.02 to acknowledge fully that
+	// 98% uptime and 100% uptime is valued the same.
+	uptimeRatio := float64(uptime) / float64(uptime+downtime)
+	if uptimeRatio > 0.98 {
+		uptimeRatio = 0.98
+	}
+	uptimeRatio += 0.02
+
+	// Cap the total amount of downtime allowed based on the total number of
+	// scans that have happened.
+	allowedDowntime := 0.03 * float64(len(host.ScanHistory))
+	if uptimeRatio < 1-allowedDowntime {
+		uptimeRatio = 1 - allowedDowntime
+	}
+
+	// Calculate the penalty for low uptime. Penalties increase extremely
+	// quickly as uptime falls away from 95%.
+	//
+	// 100% uptime = 1
+	// 98%  uptime = 1
+	// 95%  uptime = 0.83
+	// 90%  uptime = 0.26
+	// 85%  uptime = 0.03
+	// 80%  uptime = 0.001
+	// 75%  uptime = 0.00001
+	// 70%  uptime = 0.0000001
+	exp := 200 * math.Min(1-uptimeRatio, 0.30)
+	return math.Pow(uptimeRatio, exp)
+}
+
+// versionAdjustments will adjust the weight of the host according to the siad
+// version reported by the host.
+func (db *DB) versionAdjustments(host Host) float64 {
+	min := func(a, b int) int {
+		if a < b {
+			return a
+		}
+		return b
+	}
+
+	splitVersion := func(v string) (version []int, rc int) {
+		parts := strings.Split(v, "-rc")
+		for _, s := range strings.Split(parts[0], ".") {
+			n, _ := strconv.Atoi(s)
+			version = append(version, n)
+		}
+		if len(parts) == 1 { // if we don't have an rc part, we're done
+			return
+		} else if len(parts[1]) == 0 { // -rc is equivalent to -rc1 since rc0 is a full release
+			return version, 1
+		}
+
+		rc, _ = strconv.Atoi(parts[1])
+		return
+	}
+
+	versionCmp := func(a, b string) int {
+		va, rca := splitVersion(a)
+		vb, rcb := splitVersion(b)
+
+		for i := 0; i < min(len(va), len(vb)); i++ {
+			if va[i] < vb[i] {
+				return -1
+			} else if va[i] > vb[i] {
+				return 1
+			}
+		}
+
+		switch {
+		case len(va) < len(vb): // a has fewer digits than b
+			return -1
+		case len(va) > len(vb): // a has more digits than b
+			return 1
+		case rca == rcb: // length is equal and rcs are equal
+			return 0
+		case rca == 0: // a is a full release
+			return 1
+		case rcb == 0: // b is a full release
+			return -1
+		case rca > rcb:
+			return 1
+		case rca < rcb:
+			return -1
+		}
+
+		return 0
+	}
+
+	base := float64(1)
+	if versionCmp(host.Settings.Version, "2.0.0") < 0 {
+		base = math.SmallestNonzeroFloat64
+	}
+	return base
+}
+
+// Score returns the score of a host for a given allowance.
+func (db *DB) Score(pk types.PublicKey, allowance Allowance) float64 {
+	host, err := db.Host(pk)
+	if err != nil {
+		return 0
+	}
+	return db.acceptContractAdjustments(host) *
+		db.lifetimeAdjustments(host) *
+		// db.basePriceAdjustments(host) *
+		db.collateralAdjustments(host, allowance) *
+		db.durationAdjustments(host, allowance) *
+		db.interactionAdjustments(host) *
+		// db.priceAdjustments(host, allowance) *
+		// db.storageRemainingAdjustments(host, allowance) *
```

Comment:
I think we can get away with a small set of requirements instead of a full-blown `Allowance` type.
Maybe an API like:
```go
func Random(n int, reject func(Host) bool) []Host
```
`reject` can be a fairly simple function that checks whether the `WindowSize` is too small and whether the collateral/price is acceptable. That way, the hostdb doesn't need to have any concept of an allowance.